MovingBlocks/Terasology
# Record replay release

### Contains

A new feature: Record and Replay

With this, it is now possible to play the game on singleplayer while recording some events and then replay them later.

### How to test

The main menu now has 2 new buttons: One for the Recording and another one for the Replay.
Once you click on the "Record" button and go to the recording menu, every game world will appear, similar to the "select game" screen. If you select to load the game, the game will be loaded and the recording will be activated. The recording stops once the game is closed or you leave to the main menu.
When you have a recording, it will appear on the "Replay" menu. When the replay is loaded, you won't be able to make your character move, and it will take some time before the events begin to be replayed. Once the replay stops, you regain control of your character.

### Outstanding before merging

- [x] There are two bugs I am aware of and trying to solve. The first one is that when you delete a game, it will still appear in the "Record" menu. The other one is that trying to record the same world twice before closing the game glitches the replay.
- [ ] Only some types of Events are recorded (InputEvents, PlaySoundEvent, CameraTargetChangedEvent, and CharacterMoveInputEvent), so some actions won't be recorded and can cause the replays to have a weird behavior.
- [x] Lag affects the replays, so it is possible for the same replay to have different behaviors. This is I intend to improve in the next month
